modules/mqtt/destination/mqtt-destination.c

```diff
@@ -0,0 +1,266 @@
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  static gchar stats_instance[1024];
+
+  if (((LogPipe *)d)->persist_name)
+    g_snprintf(stats_instance, sizeof(stats_instance), "%s", ((LogPipe *)d)->persist_name);
+  else
+    g_snprintf(stats_instance, sizeof(stats_instance), "mqtt,%s,%s", self->address->str,
+               self->topic_name->template);
+
+  return stats_instance;
+}
+
+static const gchar *
+_format_persist_name(const LogPipe *d)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  static gchar persist_name[1024];
+
+  if (d->persist_name)
+    g_snprintf(persist_name, sizeof(persist_name), "mqtt-destination.%s", d->persist_name);
+  else
+    g_snprintf(persist_name, sizeof(persist_name), "mqtt-destination.(%s,%s)", self->address->str,
+               self->topic_name->template);
+
+  return persist_name;
+}
+
+static void
+_set_default_value(MQTTDestinationDriver *self, GlobalConfig *cfg)
+{
+  self->address       = g_string_new(DEFAULT_ADDRESS);
+
+  self->keepalive     = DEFAULT_KEEPALIVE;
+  self->clean_session = DEFAULT_CLEAN_SESSION;
+  self->qos           = DEFAULT_QOS;
+  self->message       = log_template_new(cfg, NULL);
+
+  log_template_compile(self->message, DEFAULT_MESSAGE_TEMPLATE, NULL);
+
+  log_template_options_init(&self->template_options, cfg);
+}
+
+static gboolean
+_init(LogPipe *d)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+
+  if (!log_threaded_dest_driver_init_method(d))
+    {
+      msg_error("ERROR, while init threaded dest.",
+                evt_tag_str("driver", self->super.super.super.id),
+                log_pipe_location_tag(&self->super.super.super.super));
+      return FALSE;
+    }
+
+  if (!self->topic_name)
+    {
+      msg_error("mqtt: the topic() argument is required for mqtt destinations",
+                evt_tag_str("driver", self->super.super.super.id),
+                log_pipe_location_tag(&self->super.super.super.super));
+      return FALSE;
+    }
+
+  if (self->super.batch_lines != -1 || self->super.batch_timeout != -1)
+    {
+      msg_error("The mqtt destination does not support the batching of messages, so none of the batching related parameters can be set (batch-timeout, batch-lines)",
+                evt_tag_str("driver", self->super.super.super.id),
+                log_pipe_location_tag(&self->super.super.super.super));
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static void
+_free(LogPipe *d)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+
+  if (self->fallback_topic_name)
+    g_free(self->fallback_topic_name);
+
+  log_template_unref(self->topic_name);
+
+  log_template_options_destroy(&self->template_options);
+  log_template_unref(self->message);
+
+  log_threaded_dest_driver_free(d);
+}
+
+LogDriver *
+mqtt_dd_new(GlobalConfig *cfg)
+{
+  MQTTDestinationDriver *self = g_new0(MQTTDestinationDriver, 1);
+
+  _set_default_value(self, cfg);
```

Comment:
Please move this after `log_threaded_dest_driver_init_instance()`.
That function is the base class' constructor, we should set subclass fields after that.

Reply:
solved by this comment: eef1c6a14cd64bdca3016e6fa9d8cc32b39ecf4a